### PR TITLE
Feature/music improvments

### PR DESCRIPTION
**Pull request by weeryan17**
# Pull request

 - [x] I have read and agreed to the [code of conduct](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CODE_OF_CONDUCT.md).
 - [x] I have read and complied with the [contributing guidelines](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CONTRIBUTING.md).
 - [x] What I'm implementing was assigned to me or was not being worked on and is a somewhat requested feature. For reference see our [GitHub projects](https://github.com/orgs/CascadeBot/projects).
 - [x] I have tested all of my changes.
 
## Added/Changed
Made CascadePlayer an extension if IPlayer to reduce getPlayer calls and this is really how your supposed to do it. Also its an interface now
Added CascadeLavaplayerPlayer and CascadeLavalinkPlayer for lavaplayer and lavalink respectfully.
Added EqualizerCommand for controlling the equalizer
Added EqualizerResetSubCommand for resting the equalizer to default values

## Feature
Cause the Equalizer is cool and would make a good donation perk.

### CODE REVIEW

File: src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            char[] lineChars = lineString.toCharArray();
+
+            int currentBar = 0;
+            for (int barHeight : heights) {
+                if (currentHeight <= barHeight) {
+                    int barLocation = barLocations.get(currentBar);
+                    char barChar = '░';
+                    if (currentBar == selectedBarNumber) {
+                        barChar = '█';
+                    }
```
Comment (mrjvs) [inline src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java]: Put the unicode chars into `UnicodeConstants`

File: src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                .append('\n').append(footerBuilder.toString());
+
+        return equalizerBuilder.toString();
+    }
+
+    public EmbedBuilder getEqualizerEmbed(Map<Integer, Float> bands, int currentBand, User requester, CommandContext context) {
+        String equalizer = getEqualizerString(bands, currentBand);
+
+        EmbedBuilder builder = MessagingObjects.getClearThreadLocalEmbedBuilder(requester);
+        if (bands.get(currentBand) > 0) { //TODO maybe implement color shift instead of changing color over certain value (need to talk to jvs about this)
```
Comment (mrjvs) [inline src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java]: yes, color shift would be marvelous

File: src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    public void setBand(int band, float gain) {
+        Map<Integer, Float> bands = new HashMap<>();
+        bands.put(band, gain);
+        setBands(bands);
+    }
+
+
+    //region Player methods
```
Comment (mrjvs) [inline src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java]: extra newline

File: src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
+package org.cascadebot.cascadebot.music;
+
+import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
```
Comment (mrjvs) [inline src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java]: copyright header missing?

File: src/main/java/org/cascadebot/cascadebot/music/CascadeLavaplayerPlayer.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package org.cascadebot.cascadebot.music;
+
+import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
```
Comment (mrjvs) [inline src/main/java/org/cascadebot/cascadebot/music/CascadeLavaplayerPlayer.java]: copyright header missing?

File: src/main/java/org/cascadebot/cascadebot/music/CascadePlayer.java [hunk 6]
```diff
@@ -1,27 +1,17 @@
-/*
- * Copyright (c) 2019 CascadeBot. All rights reserved.
- * Licensed under the MIT license.
- */
-
```
Comment (mrjvs) [inline src/main/java/org/cascadebot/cascadebot/music/CascadePlayer.java]: removed copyright header?

File: src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java [hunk 7]
```diff
@@ -0,0 +1,214 @@
+package org.cascadebot.cascadebot.commands.music;
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java]: Copyright notice?

File: src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+public class EqualizerCommand implements ICommandMain {
+
+    @Override
+    public void onCommand(Member sender, CommandContext context) {
+        if (!CascadeBot.INS.getMusicHandler().isLavalinkEnabled()) {
+            context.getTypedMessaging().replyDanger(context.i18n("commands.equalizer.not_lavalink"));
+            return;
+        }
+        CascadeLavalinkPlayer player = (CascadeLavalinkPlayer) context.getMusicPlayer();
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java]: Could you potentially do a check to make sure music player is an instance of the cascade one? Just to sanity check

File: src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+
+            currentBand.set(newBand);
+
+            message.editMessage(getEqualizerEmbed(player.getCurrentBands(), currentBand.get(), runner.getUser(), context).build()).override(true).queue();
+        }));
+        buttonGroup.addButton(new Button.UnicodeButton(UnicodeConstants.FORWARD_ARROW, (runner, channel, message) -> {
+            if (runner.getIdLong() != buttonGroup.getOwnerId()) {
+                return;
+            }
+            int newBand = currentBand.get() + 1;
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java]:
```suggestion
            int newBand = currentBand.incrementAndGet();
```

File: src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+            return;
+        }
+        CascadeLavalinkPlayer player = (CascadeLavalinkPlayer) context.getMusicPlayer();
+        AtomicInteger currentBand = new AtomicInteger();
+        ButtonGroup buttonGroup = new ButtonGroup(context.getUser().getIdLong(), context.getChannel().getIdLong(), context.getGuild().getIdLong());
+        buttonGroup.addButton(new Button.UnicodeButton(UnicodeConstants.BACKWARD_ARROW, (runner, channel, message) -> {
+            if (runner.getIdLong() != buttonGroup.getOwnerId()) {
+                return;
+            }
+            int newBand = currentBand.get() - 1;
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java]:
```suggestion
            int newBand = currentBand.decrementAndGet();
```

File: src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+            String lineString = "|" + StringUtils.repeat(" ", footerBuilder.length() - 1);
+            char[] lineChars = lineString.toCharArray();
+
+            int currentBar = 0;
+            for (int barHeight : heights) {
+                if (currentHeight <= barHeight) {
+                    int barLocation = barLocations.get(currentBar);
+                    char barChar = '░';
+                    if (currentBar == selectedBarNumber) {
+                        barChar = '█';
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerCommand.java]: Still not in UnicodeConstants?

File: src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerResetSubCommand.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+public class EqualizerResetSubCommand implements ISubCommand {
+
+    @Override
+    public void onCommand(Member sender, CommandContext context) {
+        if (!CascadeBot.INS.getMusicHandler().isLavalinkEnabled()) {
+            context.getTypedMessaging().replyDanger(context.i18n("commands.equalizer.not_lavalink"));
+            return;
+        }
+        Map<Integer, Float> bands = new HashMap<>();
+        for (int i = 0; i < 15; i++) {
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/commands/music/EqualizerResetSubCommand.java]: Magic number ✨Either make the `15` a variable or document it

File: src/main/java/org/cascadebot/cascadebot/music/MusicHandler.java [hunk 13]
```diff
@@ -75,7 +77,8 @@ public MusicHandler(CascadeBot instance) {
     JsonParser musicJsonParser = new JsonParser();
 
     private static JdaLavalink lavalink;
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/music/MusicHandler.java]: lavalink is still static

File: src/main/java/org/cascadebot/cascadebot/commands/music/PlayingCommand.java [hunk 14]
```diff
@@ -64,84 +64,99 @@
...
+                    CascadeBot.INS.getClient(),
+                    context.getChannel(),
+                    message,
+                    context.getGuild(),
+                    context.getData(),
+                    new String[0],
+                    runner,
+                    "equalizer",
+                    false
+            ));
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/commands/music/PlayingCommand.java]: `context.runOtherCommand`??

File: src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class CascadeLavalinkPlayer implements CascadePlayer {
+
+    @Getter
+    Map<Integer, Float> currentBands = new HashMap<>();
+
+    LavalinkPlayer lavalinkPlayer;
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java]: Access modifier!

File: src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+
+    public void setBands(Map<Integer, Float> bands) throws UnsupportedOperationException {
+        LavalinkSocket node =  lavalinkPlayer.getLink().getNode(false);
+        if (node != null) {
+            JSONObject json = new JSONObject();
+            json.put("op", "equalizer");
+            json.put("guildId", lavalinkPlayer.getLink().getGuildId());
+            JSONArray jsonArray = new JSONArray();
+            for (Map.Entry<Integer, Float> entry : bands.entrySet()) {
+                if (entry.getKey() < 0 || entry.getKey() > 14) {
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java]: Magic numbers!

File: src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+        return lavalinkPlayer.isPaused();
+    }
+
+    @Override
+    public long getTrackPosition() {
+        return lavalinkPlayer.getTrackPosition();
+    }
+
+    @Override
+    public void seekTo(long l) {
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java]:
```suggestion
    public void seekTo(long position) {
```

File: src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+        lavalinkPlayer.seekTo(l);
+    }
+
+    @Override
+    public void setVolume(int i) {
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java]:
```suggestion
    public void setVolume(int volumeLevel) {
```

File: src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+
+public class CascadeLavalinkPlayer implements CascadePlayer {
+
+    @Getter
+    private Map<Integer, Float> currentBands = new HashMap<>();
+
+    private LavalinkPlayer lavalinkPlayer;
+
+    public CascadeLavalinkPlayer(LavalinkPlayer lavalinkPlayer) {
+        for (int i = 0; i < 15; i++) {
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java]: Magic number! Use band count please

Comment (binaryoverload) [reply src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java]: Also this has been done twice, maybe make a resetBands function?

Comment (weeryan17) [reply src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java]: Thats for crating the initial local information on the bands. it doesn't actually set them server side.

Comment (binaryoverload) [reply src/main/java/org/cascadebot/cascadebot/music/CascadeLavalinkPlayer.java]: Got it

**Pull request merged**